packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,875 @@
+import {ForkName, ForkPostDeneb, ForkPostFulu, ForkPreDeneb} from "@lodestar/params";
+import {BlobIndex, ColumnIndex, SignedBeaconBlock, Slot, deneb, fulu} from "@lodestar/types";
+import {fromHex, prettyBytes, toHex, withTimeout} from "@lodestar/utils";
+import {VersionedHashes} from "../../../execution/index.js";
+import {kzgCommitmentToVersionedHash} from "../../../util/blobs.js";
+import {byteArrayEquals} from "../../../util/bytes.js";
+import {BlockInputError, BlockInputErrorCode} from "./errors.js";
+import {
+  AddBlob,
+  AddBlock,
+  AddColumn,
+  BlobMeta,
+  BlobWithSource,
+  BlockHeaderMeta,
+  BlockInputInit,
+  BlockStatus,
+  ColumnMeta,
+  ColumnWithSource,
+  CreateBlockInputMeta,
+  CustodyConfig,
+  DAData,
+  DARequirement,
+  DAStatus,
+  DAType,
+  IBlockInput,
+  LogMetaBasic,
+  LogMetaBlobs,
+  LogMetaColumns,
+  PromiseParts,
+  SourceMeta,
+} from "./types.js";
+
+export type BlockInput = BlockInputPreData | BlockInputBlobs | BlockInputColumns;
+
+export function createPromise<T>(): PromiseParts<T> {
+  let resolve!: (value: T) => void;
+  let reject!: (e: Error) => void;
+  const promise = new Promise<T>((_resolve, _reject) => {
+    resolve = _resolve;
+    reject = _reject;
+  });
+  return {
+    promise,
+    resolve,
+    reject,
+  };
+}
+
+type BlockInputState<F extends ForkName> =
+  | {
+      blockStatus: BlockStatus.MissingBlock;
+      daStatus: DAStatus.IncompleteData;
+    }
+  | {
+      blockStatus: BlockStatus.MissingBlock;
+      daStatus: DAStatus.CompleteData;
+    }
+  | {
+      blockStatus: BlockStatus.HasBlock;
+      daStatus: DAStatus.IncompleteData;
+      block: SignedBeaconBlock<F>;
+      source: SourceMeta;
+    }
+  | {
+      blockStatus: BlockStatus.HasBlock;
+      daStatus: DAStatus.CompleteData;
+      block: SignedBeaconBlock<F>;
+      source: SourceMeta;
+      timeCompleteSec: number;
+    };
+
+abstract class AbstractBlockInput<F extends ForkName = ForkName, TData extends DAData = DAData>
+  implements IBlockInput<F, TData>
+{
+  abstract type: DAType;
+  daRequirement: DARequirement;
+  timeBeginSec: number;
+
+  forkName: ForkName;
+  slot: Slot;
+  blockRootHex: string;
+  parentRootHex: string;
```

Comment:
forkName, slot and parentRootHex is not guaranteed to exist for all situations.  They need to be optionally undefined for the `unkown` block case

```suggestion
  forkName?: ForkName;
  slot?: Slot;
  blockRootHex: string;
  parentRootHex?: string;
```